man/man1/galaxy.1
.TH GALAXY1 1NEMO "6 February 2018"

.SH "NAME"
galaxy \-  3D FFT N-body code

.SH "SYNOPSIS"
\fBgalaxy\fP 

.SH "DESCRIPTION"
\fIgalaxy\fP 
is a direct descendant of the code first used by James & Sellwood (1978, MNRAS
v182, p331) with only minor refinements since.  The Poisson solver was
written exclusively by Richard James and uses the algorithm descibed in
James (1977, J Comp Phys v25, p71).  This algorithm determines the
gravitational potential of an isolated mass distribution using FFTs on a
3-D Cartesian grid without the need to remove images by doubling the grid
size in each dimension.  The motion of particles within the grid volume is
integrated forwards in time according to accelerations determined by
differencing the grid potential values.  Particles that leave the grid
volume are discarded.
.PP
Very large numbers of particles can be employed at little cost since most cpu
time is taken up by the determination of the potential on the grid.  [See
Sellwood (1997, in "Computational Astrophysics" ed Clarke & West, ASP Conf
series v123, p215) for a performance comparison with other codes.]
.PP
The version of the algorithm used here requires the individual grid cells to
be cubic, but the overall grid need not be cubic.  The FFTs supplied require
the number of mesh spaces in each direction to be (2**n + 1), where the
exponent n may be chosen independently for each coordinate direction.  As
large grids require a great deal of memory (c400 MB for a 257**3 grid), it
is recommended that the parameters set in the include file '\fBrjinclude.h\fP'
be no larger than necessary, although the code will function correctly as long
as the actual dimensions used do not exceed those set by the parameter
statement at compile time.
.PP
Time integration follows the standard 2nd order time-centered leap-frog, with
the velocities one half a time step out of synchrony with positions.  This
difference is maintained in the internally stored coordinates and is created
and can be undone, by a call to subroutine TMCENT.  For output of the particle
coordinates at a particular instant, the velocities need to be the average of
those before and after the time for the positions.
.PP
Results, in this public version, are simply the phase space coordinates of
all the particles as often as requested, which can create a very large file.
The authors therefore do not employ this scheme themselves, preferring
instead to measure and save properties of the model as the simulation
evolves.
An example of this "on the fly" analysis is provided
in the routines ICHECK and MEASURE
to determine the global integrals (energy, momentum, etc)
.PP
The grid is set up in subroutine GRDSET using parameters
read in from a short ASCII input file (\fIgalaxy.dat\fP).
The positions and velocities of the particles are read in subroutine LOADUP,
from a local file \fIgalaxy.ini\fP.
The gravitational field is determined by a call to FINDF.
The model is integrated forward by a call to STEP.
After the desired evolution is completed, the positions and velocities of
the particles are saved by a call to UNLOAD.
.SH 
The new 2014 public release of \fBgalaxy\fP contains an updated code, including
versions of SCF (see \fIscfm(1NEMO)\fP) and BHTREE (see \fBhackcode1(1NEMO)\fP).

.SH "GALAXY15"
The new release of GALAXY15 contains a fairly large number of executables that do not
confirm to the NEMO user interface. See the manual.pdf for more details, but in summary
they are:
.nf
analys
cold
compress
corrplt
dfiter
dflook
dmp2pcs
estfreq
gadget2pcs
galaxy
genplt
isotropy
merge
mkpcs
modefit
pcs2dmp
plotpft
ptest
smooth
tipsy2pcs
weed
.fi


.SH "FILES"
.nf
.ta +2i
$NEMO/src/nbody/evolve/sellwood  	old source code tree V1.3 (w/ res2snap & snap2ini)
$NEMO/usr/sellwood/GALAXY15	source code tree for GALAXY15
$NEMO/usr/sellwood//manual.pdf	The Manual
.fi
The main files associated with the run are (notice the basename 'galaxy' is fixed by the code):
.nf
.ta +2i
galaxy.dat	ASCII input: grid parameters, length and time scales
galaxy.ini	ASCII input: initial coordinates of all the particles
galaxy.lis	ASCII output: a brief summary of progress (appended)
galaxy.fin	ASCII output: final coordinates of all the particles
galaxy.res	binary output: coordinates and potentials at intervals  (appended)
galaxy.tmp	short ASCII file (deleted when closed)
galaxy.aux	large binary file (deleted when closed)
.fi
.PP
Sample \fIgalaxy.dat\fP initialization file:
.nf
 33  33  33    # number of grid cells in (x,y,z)
 15.0          # number of grid cells per length unit
 0.05          # time step length
 0.5           # time between particle outputs
 0.1           # time between integral checks
 1.00          # end time
.fi
.PP
The format of the ASCII \fIgalaxy.ini\fP (and also \fIgalaxy.fin\fP) files
is:
.nf
    Time Mass Nbody
    X_1 Y_1 Z_1 VX_1 VY_1 VZ_1
    ....
    X_n Y_n Z_n VX_n VY_n VZ_n
.fi
The \fIgalaxy.fin\fP can also be used for a restart. Note that the timestep
is one more then the last requested time, to prevent that the 
\fIgalaxy.res\fP file will contain a datadump on the restart timestep twice.

.SH "COMPILATION"
Note that the maximum grid size and maximum amount of particles are hardcoded
during compilation. See comments in \fBrjinclude.h\fP how to change the
maximum grid size and the parameter \fBmbod\fP in
\fBgalaxy.f\fP for the maximum particle number that can be used.
.PP
The new release V15 does not depend on commercial (e.g. NAG) software, and
can be compiled using open source tools and libraries. PGPLOT must be
installed independantly.

.SH "RES2SNAP"
The following shell script, given as an example, will convert a particular
snapshot from a binary "galaxy.res" file into a NEMO \fIsnapshot(5NEMO)\fP
format.
.PP
.nf
#! /bin/csh -f
if ($#argv != 2) then
  echo Usage: res2snap FILE N
  echo Converts the N-th snapshot from FILE to NEMOs snapshot format
  exit 0
endif
#   set command line parameters
set file=$1
set n=$2
#   get header info 
set tsnap=`unfio $file "$n*2-1" float | awk '{if (NR==1) print $1}'`
set nbody=`unfio $file "$n*2-1" int   | awk '{if (NR==2) print $1}'`
#   dump data and convert to snapshot
unfio $file "$n*2" float maxbuf=100000 |\\
   tabtos - ${file:r}.$n.snap "" pos,vel,phi options=wrap times=$tsnap nbody=$nbody
.fi

.SH "BENCHMARKS"
1000 particles on a 32**3 cube integrating for 20 steps took 4.77\" on
a sparc Ultra-2. (1997?)
.PP
2021 update: 200 steps on an i5-1135G7 took 3.7\" using the default gfortran, only a factor
13 speedup from the 2004 results. However, compiling with flang (FFLAGS=-O3), 200 steps took
0.22\", a remarkable speedup of 16 over gfortran.

.SH "SEE ALSO"
rungalaxy(1NEMO), unfio(1NEMO), snapshot(5NEMO)
.PP
.nf
James & Sellwood (1978, MNRAS v182, p331) 
James (1977, J Comp Phys v25, p71).  
Sellwood (1997, in "Computational Astrophysics" ed Clarke & West, ASP Conf series v123, p215) 
http://www.physics.rutgers.edu/galaxy   (full 2014 version; V1.5 and later)
.fi

.SH "ADS"
@ads 2014arXiv1406.6606S

.SH "AUTHOR"
Jerry Sellwood

.SH "HISTORY"
.nf
.ta +1i +4i
9-jun-97	V1.0  Sellwood public release/adopted for NEMO  	JS/PJT
24-jun-97	V1.1  added ICHECK/MEASURE; dtlog to galaxy.dat 	JS
18-mar-04	fixed bad usage line; refer to rungalaxy now	PJT
8-mar-06	V1.3  now installs by default into NEMOBIN	PJT
26-jun-2014	notes on the official full public release	PJT
10-mar-2017     notes on the new V15 release	PJT
6-feb-2018	notes on the new V1.5.2 release	PJT
.fi


src/nbody/evolve/sellwood/code1/galaxy.f
      program galaxy
c program for simulation of an isolated galaxy written by:
c
c                Jerry Sellwood and Richard James
c
c It is a direct descendant of that first used by James & Sellwood (1978, MNRAS
c   v182, p331) with only minor refinements since.  The Poisson solver was
c   written exclusively by Richard James and uses the algorithm descibed in
c   James (1977, J Comp Phys v25, p71).  This algorithm determines the
c   gravitational potential of an isolated mass distribution using FFTs on a
c   3-D Cartesian grid without the need to remove images by doubling the grid
c   size in each dimension.  The motion of particles within the grid volume is
c   integrated forwards in time according to accelerations determined by
c   differencing the grid potential values.  Particles that leave the grid
c   volume are discarded.
c
c Very large numbers of particles can be employed at little cost since most cpu
c   time is taken up by the determination of the potential on the grid.  [See
c   Sellwood (1997, in "Computational Astrophysics" eds Clarke & West, ASP Conf
c   series v123, p215) for a performance comparison with other codes.]
c
c The version of the algorithm used here requires the individual grid cells to
c   be cubic, but the overall grid need not be cubic.  The FFTs supplied
c   require the number of mesh spaces in each direction to be (2**n + 1), where
c   the exponent n may be chosen independently for each coordinate direction.
c   As large grids require a great deal of memory (c400 MB for a 257**3 grid),
c   it is recommended that the parameters set in the include file 'rjinclude.h'
c   be no larger than necessary, although the code will function correctly as
c   long as the actual dimensions used do not exceed those set by the parameter
c   statement at compile time.
c
c Time integration follows the standard 2nd order time-centered leap-frog, with
c   the velocities one half a time step out of synchrony with positions.  This
c   difference is maintained in the internally stored coordinates and is
c   created, and can be undone, by a call to subroutine TMCENT.  For output of
c   the particle coordinates at a particular instant, the velocities need to be
c   the average of those before and after the time for the positions.
c
c Results, in this public version, are simply the phase space coordinates of
c   all the particles as often as requested, which can create a very large
c   file.  The authors therefore do not employ this scheme themselves,
c   preferring instead to measure and save properties of the model as the
c   simulation evolves.  An example of this "on the fly" analysis is provided
c   in the procedure to determine the global integrals (energy, momentum, etc).
c
c The grid is set up in subroutine GRDSET using data read in from a short ASCII
c   input file (galaxy.dat)
c The positions and velocities of the particles are read in subroutine LOADUP
c The gravitational field is determined by a call to FINDF
c The model is integrated forward by a call to STEP
c After the desired evolution is completed, the positions and velocities of
c   the particles are saved by a call to UNLOAD
c
c The main files associated with the run are:
c   galaxy.dat - ASCII input: grid parameters, length and time scales
c   galaxy.lis - ASCII output: a brief summary of progress (appended)
c   galaxy.ini - ASCII input: initial coordinates of all the particles
c   galaxy.fin - ASCII output: final coordinates of all the particles
c   galaxy.res - binary output: coords and potentials at intervals (appended)
c
c There are also two scratch files:
c   galaxy.aux - short ASCII file (deleted when closed)
c   galaxy.tmp - large binary file (deleted when closed)
c
c common block
c
      include 'admin.h'
c
c make sure block data segment is linked
      external blkdat
c
c local variables
      integer i, ilast, j, l
      real tend

c say hello and remind the user of the max- parameters
      print *, 'galaxy V1.3 '
      print *, 'Maximum number of particles (mbuff): ',mbuff
c
c store the sizes of the main common arrays
      call dimens
c read galaxy.dat file (ASCII input) and se up grid
      call grdset
      read( ni, * )tend
      close( ni )
c create Green's function for Poisson solver
      call greenm( .false. )
c read the galaxy.ini file
      call loadup
      print *, 'Run (re)started at time', ts * real( istep )
      ilast = nint( tend / ts ) + 1
      print *, '  and will stop after the step at time ',
     +         ts * real( ilast )
      if( istep .ge. ilast )call crash( 'GALAXY',
     +                                  'End time already passed' )
c assign masses for the current particle positions
      call masset
c find gravitational field
      call findf
c back up velocities one half step
      call tmcent( .true. )
c main time-step sequence
      do while ( istep .lt. ilast )
        print *, 'Starting step', istep
        write( no, * )'Starting step', istep
c move particles
        call step
c increment step number
        istep = istep + 1
c calculate new gravitational field
        call findf
      end do
      write( no, * )'Run stopped at time', ts * real( istep )
c step forward velocities one half step
      call tmcent( .false. )
c update the .dmp file
      call unload
      stop
      end

      subroutine dimens
c written by Jerry Sellwood for the galaxy simulation code
c
c Sets the dimensions of the two adjustable arrays in the common blocks
c   / grids / and / ptcls /  They are dimensioned to have size 1 in every
c   other routine except for this in order to avoid having to recompile
c   the entire library whenever the size of the problem is changed
c
c The actual sizes of these arrays are stored in the first variable in
c   each of these blocks
c
      integer mbod, par1, par2
      parameter ( mbod = 10000 )
c
      include 'rjinclude.h'
c
c common blocks
c
c size needs to be large enough to store all the particles and pointers
      parameter ( par1 = 8 * mbod )
      integer lptcls
      real ptcls( par1 )
      common / ptcls / lptcls, ptcls
c
c size needs to be large enough to store coordinates plus potential
      real coor( 7 * mbod )
      common / results / coor
c
c size needs to be two whole meshes (for masses and potentials) plus 6 planes
c (one pair for each acceleration component)
      parameter ( par2 = 2 * rjn1 * rjn2 * rjn3 + 6 * rjn2 * rjn3 )
      integer lgrids
      real grids( par2 )
      common / grids / lgrids, grids

c
c say hello and remind user to the size
      print *, 'Maximum gridsize: ',rjn1,rjn2,rjn3
c
c store sizes in common blocks
      lptcls = par1
      lgrids = par2
      return
      end
